### ISSUE TITLE
API and impl jar should have same module-name

### ISSUE THREAD

**Issue by lukasj** [eclipse-ee4j/mail#409]
**Describe the bug**
missing module-name/automatic module name in an API jar (thus reliance on the filename based module resolution) makes transition to JDK11/JPMS not so straightforward

**To Reproduce**
* to make sure compilation gets just API and runtime/tests do get also an implementation ('test' scope then needs to explicitly exclude API) on JDK 8 one can use following setup in a maven project:
```
    <dependencies>
        <dependency>
            <groupId>jakarta.mail</groupId>
            <artifactId>jakarta.mail-api</artifactId>
            <version>1.6.4</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>com.sun.mail</groupId>
            <artifactId>jakarta.mail</artifactId>
            <version>1.6.4</version>
            <scope>runtime</scope> <!-- 'test' might be more common here -->
        </dependency>
    </dependencies>
```
* to move to JDK11, let's create module-info in a project having simple ```Hello world``` main class, so it compiles properly as module:
```
module org.errors.bug {
    requires jakarta.mail.api;
}
```
* run the project:
```
ljungman-mac:target lukas$ java -p jakarta.mail-1.6.4.jar:jakarta.activation-1.2.1.jar:mavenproject14-1.0-SNAPSHOT.jar -m org.errors.bug
Error occurred during initialization of boot layer
java.lang.module.FindException: Module jakarta.mail.api not found, required by org.errors.bug
```
* the only way to make this work properly is to have direct compile dependency on the jakarta mail implementation artifact and to have following in the module info:
```
module org.errors.bug {
    requires jakarta.mail;
}
```

**Expected behavior**
definition of the module name of the jakarta mail implementation should allow me to be able to easily switch between API jar/implementation and/or implementation from other vendor (who may not provide APIs directly within his impl jar), so the best would be if the module name of both artifacts is the same

**Comment by bshannon** [comment]
Module names are never defined with ".api" - the module name **is** the API.

If there are separate API modules and implementation modules, the API module
should have the standard name of the API, and the implementation module should
have a name corresponding to the implementation entity, not the specification.

Thus, the API module is jakarta.mail and the implementation module is com.sun.mail.

If the API and implementation are combined in one module, with only the standard
API exposed, the module should have the name of the standard API, e.g.,
"jakarta.mail" (although we have no such module here).

Note that there might be many modules from different implementors with the same
standard module name.  Only the Maven artifact coordinates will distinguish one
implementation from another.

**Comment by lukasj** [comment]
> Module names are never defined with ".api"

```
ljungman-mac:Downloads lukas$ jar --describe-module --file jakarta.mail-api-2.0.0-RC1.jar 
No module descriptor found. Derived automatic module.

jakarta.mail.api@2.0.0-RC1 automatic
requires java.base mandated
contains jakarta.mail
contains jakarta.mail.event
contains jakarta.mail.internet
contains jakarta.mail.search
contains jakarta.mail.util
```

While it is true that no module name is explicitly defined (= no `.api` is mentioned anywhere), automatic module name resolution algorithm resolves the name from the jar file name to `jakarta.mail.api` which looks to me like contradiction to `Module names are never defined with ".api"` statement.

**Comment by bshannon** [comment]
Which, as you know, is the reason to add the Automatic-Module-Name manifest
header, because the default name algorithm has many flaws.

**Comment by lukasj** [comment]
The point is there is no Automatic-module-name header in api jar’s manifest, at least I don’t see it in 2.0.0-rc1

**Comment by bshannon** [comment]
Ah, you're right.  Fixed in 2.0.0-RC3.
